src/main/kotlin/edu/cornell/cs/apl/viaduct/selection/GreedySelection.kt

```diff
@@ -0,0 +1,65 @@
+package edu.cornell.cs.apl.viaduct.selection
+
+import edu.cornell.cs.apl.attributes.attribute
+import edu.cornell.cs.apl.viaduct.analysis.InformationFlowAnalysis
+import edu.cornell.cs.apl.viaduct.analysis.NameAnalysis
+import edu.cornell.cs.apl.viaduct.protocols.Local
+import edu.cornell.cs.apl.viaduct.syntax.HostTrustConfiguration
+import edu.cornell.cs.apl.viaduct.syntax.Protocol
+import edu.cornell.cs.apl.viaduct.syntax.Variable
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.DeclarationNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.InputNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.LetNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.Node
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.ProcessDeclarationNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.QueryNode
+
+class GreedySelection(val selector: ProtocolSelector, val protocolSort: (Protocol) -> Int) {
+    fun select(
+        processDeclaration: ProcessDeclarationNode,
+        nameAnalysis: NameAnalysis,
+        informationFlowAnalysis: InformationFlowAnalysis
+    ): (Variable) -> Protocol {
+        val hostTrustConfiguration = HostTrustConfiguration(nameAnalysis.tree.root)
+        var assignment: MutableMap<Variable, Protocol> = mutableMapOf()
+        val protocolSelection = object {
+            private val LetNode.possibleProtocols: Set<Protocol> by attribute {
+                when (value) {
+                    is InputNode ->
+                        setOf(Local(value.host.value))
+                    is QueryNode -> nameAnalysis.declaration(value).possibleProtocols
+                    else ->
+                        selector.selectLet(assignment, this)
+                }
+            }
+
+            private val DeclarationNode.possibleProtocols: Set<Protocol> by attribute {
+                selector.selectDeclaration(assignment, this)
+            }
+
+            fun possibleProtocols(node: LetNode) = node.possibleProtocols
+
+            fun possibleProtocols(node: DeclarationNode) = node.possibleProtocols
+        }
+
+        fun traverse(node: Node) {
```

Comment:
This does no backtracking. Isn't that an issue? I know it's called greedy search, but I think that just means it's not globally optimizing, it's picking the first solution it finds. This may not find a solution even if there is one...

Reply:
It should have a different name perhaps? It is the same algorithm as the dumb protocol selection from before